File: backend/src/modules/order/application/use-cases/create-order.usecase.ts
import { Injectable, Inject } from '@nestjs/common';
import { ORDER_REPOSITORY } from '@modules/order/domain/repositories/order.repository';
import type { OrderRepository } from '@modules/order/domain/repositories/order.repository';
import { UseCase } from '@shared/application/usecase.interface';
import type { OutputPort } from '@shared/application/contracts/output-port.interface';
import { CreateOrderRequest } from '@modules/order/application/contracts/requests/create-order.request';
import { CreateOrderResponse } from '@modules/order/application/contracts/responses/create-order.response';
import { OrderEntity } from '@modules/order/domain/order.entity';
import { OrderItem } from '@modules/order/domain';

@Injectable()
export class CreateOrderUseCase implements UseCase<CreateOrderRequest> {
  constructor(
    private readonly outputPort: OutputPort<CreateOrderResponse>,
    @Inject(ORDER_REPOSITORY) private readonly orders: OrderRepository,
  ) {}

  async execute(input: CreateOrderRequest): Promise<void> {
    const priceBreakdown = OrderEntity.calculateTotalWithTax(
      input.items.map((item) => ({
        name: item.name,
        price: item.price,
        quantity: item.quantity,
      })) as OrderItem[],
    );

    const order = await this.orders.createOrder(
      input.customerName,
      input.items,
      priceBreakdown.total,
    );

    this.outputPort.present({
      id: order.id,
      customerName: order.customerName,
      totalAmount: order.totalAmount,
      status: order.status as string,
      items: order.items,
      estimatedReadyTime: order.estimatedReadyTime,
      createdAt: order.createdAt,
    });
  }
}

File: frontend/lib/features/order/domain/usecases/create_order_usecase.dart
import '../../../../core/domain/usecase.dart';
import '../../../../core/utils/result.dart';
import '../../domain/repositories/order_repository.dart';

/// Use case: Create Order
class CreateOrderUseCase
    implements UseCase<CreateOrderParams, Result<Map<String, dynamic>>> {
  final OrderRepository _repository;

  CreateOrderUseCase(this._repository);

  @override
  Future<Result<Map<String, dynamic>>> call(CreateOrderParams params) {
    return _repository.createOrder(
      customerName: params.customerName,
      items: params.items,
    );
  }
}

class CreateOrderParams {
  final String customerName;
  final List<Map<String, dynamic>> items;

  const CreateOrderParams({required this.customerName, required this.items});
}
